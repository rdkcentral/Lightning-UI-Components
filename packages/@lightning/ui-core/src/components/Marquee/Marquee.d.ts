import lng from '@lightningjs/core';
import Base from '../Base';
import type { StylePartial } from '../../types/lui';

export type MarqueeStyle = {
  fadeW: number;
  offset: number;
  shouldSmooth: boolean;
};

export default class Marquee extends Base {
  autostart?: boolean;
  centerAlign?: boolean;
  contentTexture?: lng.Texture;
  delay?: number;
  repeat?: number;
  title?: string;
  get style(): MarqueeStyle;
  set style(v: StylePartial<MarqueeStyle>);

  // tags
  get _ContentClipper(): lng.Component;
  get _ContentBox(): lng.Component;
  get _Content(): lng.Component;
  get _ContentLoopTexture(): lng.Component;
}
