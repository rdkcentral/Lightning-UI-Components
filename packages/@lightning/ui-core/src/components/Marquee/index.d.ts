import Marquee, { MarqueeStyle } from './Marquee';

export { Marquee as default, MarqueeStyle };
