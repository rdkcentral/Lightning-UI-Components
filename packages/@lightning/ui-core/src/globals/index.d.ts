export { default as context } from './context';
