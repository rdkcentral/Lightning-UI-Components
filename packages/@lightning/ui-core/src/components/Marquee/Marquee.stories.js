import lng from '@lightningjs/core';

import MarqueeComponent from '.';
import InlineContent from '../InlineContent/index.js';
import mdx from './Marquee.mdx';
import { context } from '../../globals/index.js';
import { CATEGORIES } from 'lightning-ui-docs';

export default {
  title: `${CATEGORIES[256]}/Marquee`,
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const Marquee = args =>
  class Marquee extends lng.Component {
    static _template() {
      return {
        Marquee: {
          type: MarqueeComponent,
          w: 350,
          h: 180,
          title: {
            text: args.title,
            ...context.theme.typography.body1,
            maxLines: 1
          }
        }
      };
    }
  };

Marquee.args = {
  title: 'This is a scrolling Marquee component for long text',
  autoStart: true,
  centerAlign: false,
  repeat: -1,
  delay: 1.5
};

Marquee.argTypes = {
  title: {
    control: 'text',
    description: 'An object matching Lightning text object'
  },
  autoStart: {
    control: { type: 'boolean' },
    description: 'Start scrolling on initiation'
  },
  color: {
    name: 'color',
    control: { type: 'color' },
    description: 'Set color of text'
  },
  repeat: {
    control: { type: 'number', min: -1 },
    description: 'Number of times to repeat scrolling'
  },
  delay: {
    control: { type: 'number', min: 1.5 },
    description: 'Delay before scrolling starts'
  },
  centerAlign: {
    control: { type: 'boolean' },
    description:
      'Will center the entire texture if the width of the content is less than the container'
  }
};

Marquee.parameters = {
  argActions: {}
};

export const CenteredText = () =>
  class CenteredText extends lng.Component {
    static _template() {
      return {
        Marquee: {
          type: MarqueeComponent,
          w: 400,
          h: 50,
          title: {
            text: 'No scroll center',
            textAlign: 'center'
          },
          autoStart: true,
          delay: 1
        },
        InlineContent: {
          type: InlineContent,
          x: 200,
          alpha: 0.001,
          justify: 'flex-start',
          content: [
            'Centered',
            {
              icon: 'https://upload.wikimedia.org/wikipedia/commons/b/b6/Tomato-Torrent-Icon.png',
              title: 'Rotten Tomatoes rating'
            },
            ' InlineContent'
          ],
          rtt: true
        },
        MarqueeInlineContent: {
          type: MarqueeComponent,
          y: 100,
          w: 400,
          h: 180,
          centerAlign: true,
          alpha: 0,
          autoStart: true,
          delay: 1
        }
      };
    }

    _init() {
      this.tag('MarqueeInlineContent').contentTexture =
        this.tag('InlineContent').getTexture();
      this.tag('MarqueeInlineContent').alpha = 1;
    }
  };

CenteredText.args = {};
CenteredText.argTypes = {};
CenteredText.parameters = {
  argActions: {}
};
