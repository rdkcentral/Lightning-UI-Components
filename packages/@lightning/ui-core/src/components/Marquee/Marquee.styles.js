export const base = theme => ({
  fadeW: theme.layout.gutterX.lg,
  offset: theme.spacer.xxl,
  shouldSmooth: false,
  textStyle: {
    ...theme.typography.body1,
    textAlign: 'center'
  }
});
