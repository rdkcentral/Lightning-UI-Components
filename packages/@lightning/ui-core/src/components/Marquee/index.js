export { default as default } from './Marquee.js';
