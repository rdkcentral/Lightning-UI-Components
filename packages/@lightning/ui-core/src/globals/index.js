export { default as context } from './context/index.js';
export { updateManager } from './global-update-manager/GlobalUpdateManager.js';
export { default as pool } from './pool/index.js';
