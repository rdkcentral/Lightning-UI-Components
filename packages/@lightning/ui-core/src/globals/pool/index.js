/*
Create a Lightning Object pool and then get the component from the pool.

By default we create two objects, with expectations that one is used at a time

Learn more about object pools
https://egghead.io/blog/object-pool-design-pattern
*/

const pools = new Map();

function get(key) {
  if (pools.has(key)) {
    const pool = pools.get(key);
    let index = pool.index;

    if (index >= pool.components.length) {
      index = 0;
    }
    pool.index = index + 1;

    return pool.components[index];
  }

  return false;
}

function clear() {
  pools.clear();
}

function create({ name, component, stage, size = 2 }) {
  if (pools.has(name)) {
    return get(name);
  }

  component = component || { type: name };
  const components = [];
  for (let i = 0; i < size; i++) {
    components.push(stage.c(component));
  }

  pools.set(name, {
    index: 0,
    components
  });

  return get(name);
}

export default {
  get,
  create,
  clear
};
