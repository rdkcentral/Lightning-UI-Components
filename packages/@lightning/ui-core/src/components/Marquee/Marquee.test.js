import { makeCreateComponent } from '@lightning/ui-test-utils';
import Marquee from './index.js';
import { jest } from '@jest/globals';

const createMarquee = makeCreateComponent(Marquee, {
  w: 350,
  h: 50,
  title: {
    text: 'This is a scrolling Marquee component for long text',
    lineHeight: 50,
    maxLines: 1
  },
  delay: 0
});

describe('Marquee', () => {
  let marquee, testRenderer;

  beforeEach(() => {
    [marquee, testRenderer] = createMarquee();
  });

  afterEach(() => {
    marquee = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('sets the announce string to the label text', () => {
    const text = 'Scrolling Text';
    marquee.title = { text };
    testRenderer.forceAllUpdates();
    expect(marquee.announce).toBe(text);
  });

  describe('scrolling', () => {
    it('does not scroll when width is big enough', () => {
      [marquee, testRenderer] = createMarquee({
        autoStart: true,
        text: {
          textAlign: 'center'
        }
      });
      marquee.startScrolling();
      expect(marquee._scrolling).toBe(false);
      [marquee, testRenderer] = createMarquee({
        autoStart: true,
        centerAlign: true
      });
      marquee.startScrolling();
      expect(marquee.centerAlign).toBe(true);
      expect(marquee._shouldCenter()).toBe(true);
      expect(marquee._scrolling).toBe(false);
    });

    it('with autoStart property', done => {
      [marquee, testRenderer] = createMarquee({
        autoStart: true
      });
      // Canvas mock doesnt render text, just uses length
      // https://github.com/hustcc/jest-canvas-mock/issues/67
      Object.defineProperty(marquee, '_textRenderedW', {
        get: jest.fn(() => 500)
      });
      marquee.startScrolling();
      setTimeout(() => {
        expect(marquee._scrollAnimation.isPlaying()).toBe(true);
        done();
      }, 50);
    });

    it('handles txLoaded', () => {
      const [marquee] = createMarquee({ w: 300, h: 300, title: '' });
      marquee.startScrolling();
      marquee._Content.emit('txLoaded', 'loaded');
      expect(marquee._scrolling).toBe(false);
    });

    it('#startScrolling', done => {
      Object.defineProperty(marquee, '_textRenderedW', {
        get: jest.fn(() => 500)
      });
      expect(marquee._scrollAnimation).toBe(undefined);
      marquee.startScrolling();
      setTimeout(() => {
        expect(marquee._scrollAnimation.isPlaying()).toBe(true);
        done();
      }, 50);
    });

    it('#startScrolling waits until rendered', done => {
      const renderWMock = jest.fn();
      renderWMock.mockReturnValueOnce(0).mockReturnValue(500);
      Object.defineProperty(marquee, '_textRenderedW', {
        get: renderWMock
      });
      marquee.startScrolling();
      setTimeout(() => {
        expect(marquee._scrollAnimation.isPlaying()).toBe(true);
        done();
      }, 50);
    });

    it('#startScrolling centers text with enough width', () => {
      [marquee, testRenderer] = createMarquee({
        title: {
          textAlign: 'center'
        }
      });
      marquee.startScrolling();
      expect(marquee._Content.text.textAlign).toBe('center');
      expect(marquee._scrolling).toBe(false);
    });

    it('#stopScrolling', done => {
      Object.defineProperty(marquee, '_textRenderedW', {
        get: jest.fn(() => 500)
      });
      marquee.startScrolling();
      expect(marquee._scrolling).toBe(true);
      setTimeout(() => {
        marquee.stopScrolling();
      }, 50);
      setTimeout(() => {
        expect(marquee._scrolling).toBe(false);
        done();
      }, 55);
    });
  });

  describe('#title', () => {
    it('updates the title', () => {
      marquee.title = {
        text: 'Hello There'
      };
      testRenderer.update();
      expect(marquee.title.text).toEqual('Hello There');
    });
  });

  describe('#color', () => {
    it('updates the text color', () => {
      const color = 0xff1c27bc;
      marquee.color = color;
      testRenderer.update();
      const tree = testRenderer.toJSON(2);
      expect(tree).toMatchSnapshot();
      expect(marquee.color).toBe(color);
    });
  });

  describe('#contentTexture', () => {
    it('updates the texture', () => {
      marquee.contentTexture = marquee.getTexture();
      testRenderer.update();
      expect(marquee.contentTexture.constructor.name).toEqual('TextureSource');
    });
  });

  describe('should position contentBox', () => {
    it('updates the texture', done => {
      [marquee, testRenderer] = createMarquee({
        w: 300,
        h: 300,
        centerAlign: true
      });
      marquee._componentStyle.shouldSmooth = true;
      marquee._centerTexture();
      setTimeout(() => {
        expect(marquee._ContentBox.x).toBe(124.5);
        done();
      }, 50);
    });
  });

  describe('should center', () => {
    it('updates the texture', () => {
      [marquee, testRenderer] = createMarquee({
        w: 300,
        h: 300,
        centerAlign: false
      });
      marquee._Content.text = 'marque is scrolling';
      marquee._Content.text.textAlign = 'center';
      expect(marquee._shouldCenter()).toBe(true);
    });
  });

  describe('should clip', () => {
    it('updates the texture', () => {
      [marquee, testRenderer] = createMarquee({
        w: 0,
        h: 300,
        shouldClip: true
      });
      expect(marquee._ContentClipper.w).toBe(0);
    });
  });

  describe('should update delay and repeat', () => {
    it('updates the texture', () => {
      [marquee, testRenderer] = createMarquee({
        w: 300,
        h: 300,
        shouldClip: true,
        delay: NaN,
        repeat: 5
      });
      marquee._updateAnimation();
      expect(marquee.animation().element.repeat).toBe(marquee.repeat);
      expect(marquee._scrollAnimation._settings.delay).toBe(1.5);
    });
  });

  describe('#shouldSmooth', () => {
    it('should transition the content box x position when centering', done => {
      [marquee, testRenderer] = createMarquee({
        autoStart: true,
        shouldSmooth: true,
        centerAlign: true,
        w: 350,
        h: 50,
        title: {
          text: 'Short text',
          lineHeight: 50,
          maxLines: 1
        }
      });
      Object.defineProperty(marquee, '_textRenderedW', {
        get: jest.fn(() => 100)
      });
      expect(marquee.shouldSmooth).toBe(true);
      setTimeout(() => {
        expect(marquee._scrolling).toBe(false);
        done();
      }, 50);
    });
  });
});
