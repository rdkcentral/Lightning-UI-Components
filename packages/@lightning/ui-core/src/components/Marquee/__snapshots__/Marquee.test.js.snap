// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Marquee #color updates the text color 1`] = `
{
  "Component": {
    "ContentClipper": {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": [],
      "children": {
        "ContentBox": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": {
            "Content": {
              "active": true,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Content",
              "renderOfScreen": undefined,
              "renderToTexture": true,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": [
                "Content",
              ],
              "texture": {
                "fontFace": "Arial",
                "fontSize": 25,
                "lineHeight": 50,
                "maxLines": 1,
                "text": "This is a scrolling Marquee component for long text",
                "textBaseline": "bottom",
                "type": "TextTexture",
                "verticalAlign": "middle",
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
            "ContentLoopTexture": {
              "active": true,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "ContentLoopTexture",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": [
                "ContentLoopTexture",
              ],
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "ContentBox",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": [
            "ContentBox",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 50,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "ContentClipper",
      "renderOfScreen": undefined,
      "renderToTexture": true,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "ContentClipper",
      ],
      "visible": true,
      "w": 390,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;

exports[`Marquee renders 1`] = `
{
  "Component": {
    "ContentClipper": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": [],
      "children": {
        "ContentBox": {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": {
            "Content": {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Content",
              "renderOfScreen": undefined,
              "renderToTexture": true,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": [
                "Content",
              ],
              "texture": {
                "fontFace": "Arial",
                "fontSize": 25,
                "lineHeight": 50,
                "maxLines": 1,
                "text": "This is a scrolling Marquee component for long text",
                "textBaseline": "bottom",
                "type": "TextTexture",
                "verticalAlign": "middle",
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
            "ContentLoopTexture": {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "ContentLoopTexture",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": [
                "ContentLoopTexture",
              ],
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "ContentBox",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": [
            "ContentBox",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "ContentClipper",
      "renderOfScreen": undefined,
      "renderToTexture": true,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "ContentClipper",
      ],
      "visible": true,
      "w": 390,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
